ui/src/notebooks/context/notebook.tsx

```diff
@@ -0,0 +1,78 @@
+import React, {FC, useState} from 'react'
+
+// TODO make this polymorphic to mimic the self registration
+// of pipe stages
+export type Pipe = any
+
+export interface NotebookContextType {
+  id: string
+  pipes: Pipe[]
+  addPipe: (pipe: Pipe) => void
+  updatePipe: (idx: number, pipe: Pipe) => void
+  removePipe: (idx: number) => void
+}
+
+export const DEFAULT_CONTEXT = {
+  id: 'new',
+  pipes: [],
+  addPipe: () => {},
+  updatePipe: () => {},
+  removePipe: () => {},
+}
+
+// NOTE: this just loads some test data for exploring the
+// interactions between types. Make sure it's never true
+// during the pull review process
+const TEST_MODE = false
+if (TEST_MODE) {
+  const TEST_NOTEBOOK = require('src/notesbooks/context/notebook.mock.json')
+  DEFAULT_CONTEXT.id = TEST_NOTEBOOK.id
+  DEFAULT_CONTEXT.pipes = TEST_NOTEBOOK.pipes
+}
+
+export const NotebookContext = React.createContext<NotebookContextType>(
+  DEFAULT_CONTEXT
+)
+
+export const NotebookProvider: FC = ({children}) => {
+  const [id] = useState(DEFAULT_CONTEXT.id)
+  const [pipes, setPipes] = useState(DEFAULT_CONTEXT.pipes)
+
+  function addPipe(pipe: Pipe) {
+    setPipes(pipes => {
+      pipes.push(pipe)
+      return pipes.slice()
```

Comment:
whats our preferred method of doing this? we've also seen this form
```ts
[...pipes] 
```

Reply:
i have no preference, but considering that it's not standardized across the code base, that if we're gonna try to make a choice about this, it not be made here, but opened up to the group and captured in a style guide that we can maintain / distribute / refer to in moments like this.